go/common/crypto/signature/signers/ledger/ledger_signer.go

```diff
@@ -0,0 +1,85 @@
+package ledger
+
+import (
+	"encoding/binary"
+
+	"github.com/ethereum/go-ethereum/accounts"
+	"github.com/oasislabs/ledger-go"
+	"github.com/oasislabs/oasis-core/go/common/crypto/signature"
+)
+
+const (
+	ETH_CLA        byte = 0xE0
+	ETH_INS_CONFIG byte = 0x06
+	ETH_INS_DERIVE byte = 0x02
+	ETH_INS_SIGN   byte = 0x04
+)
+
+type Signer struct {
+	device *ledger_go.Ledger
+}
+
+// NewSigner will attempt to initialize a Ledger Signer by running
+// FindLedger. This will connect to the first Ledger device found.
+func NewSigner() (*Signer, error) {
+	device, err := ledger_go.FindLedger()
+	if err != nil {
+		return nil, err
+	}
+	return &Signer{device}, nil
+}
+
+func (s *Signer) Public() signature.PublicKey {
+	derivationPath := accounts.DefaultRootDerivationPath
+	path := make([]byte, 1+4*len(derivationPath))
+	path[0] = byte(len(derivationPath))
+	for i, component := range derivationPath {
+		binary.BigEndian.PutUint32(path[1+4*i:], component)
+	}
+
+	header := []byte{ETH_CLA, ETH_INS_DERIVE, 0, 0, byte(len(path))}
+	message := append(header, path...)
+
+	response, err := s.device.Exchange(message)
+	if err != nil {
+		return nil
+	}
+	pkLength := int(response[0])
+	addrLength := int(response[pkLength+1])
+	return response[pkLength+2 : pkLength+2+addrLength]
+}
+
+func (s *Signer) Sign(tx []byte) ([]byte, error) {
+	derivationPath := accounts.DefaultRootDerivationPath
```

Comment:
So every single signer created with the same hardware token will get the same keypair?

Our signers are role-segregated for a reason, and are meant to be instantiated via the `signature.SignerFactory` interface.  On the off chance a user wants to regenerate/rotate their key, this appears to not support that.

Reply:
So are you saying that the design should be that there are multiple devices connected to the machine, and each are assigned a role? Each device either generates or loads a 24 word seed to derive its keys, and each device only stores a single master seed at a time. So to answer your first question, yes two devices with the same master seed will get the same keypair. If we want to have role-segregated signers and still work for a single device, we can have the roles map to different derivation paths.